### PR TITLE
perf(switch): Cache peerConfig to prevent allocating more memory.

### PR DESCRIPTION
**Pull request by itsdevbear**
Allocating a ton of memory for no reason.

image

#go-api-breaking

---

#### PR checklist

- [ ] ~~Tests written/updated~~
- [ ] ~~Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)~~
- [ ] ~~Updated relevant documentation (`docs/` or `spec/`) and code comments~~
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

**Review by melekes — APPROVED** [review]
Thanks @itsdevbear ❤️

**Comment by itsdevbear** [comment]
@melekes fixed, I don't see why this change is breaking though, mistake on tagging?

**Comment by melekes** [comment]
```go
	Accept(config *peerConfig) (Peer, error)

	// Dial connects to the Peer for the address.
	Dial(addr NetAddress, config *peerConfig) (Peer, error)
```

If I am not mistaken, this makes the PR Go API breaking.

**Comment by melekes** [comment]
```
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 518 [running]:
testing.tRunner.func1.2({0x126f820, 0x15d0170})
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1634 +0x6b6
panic({0x126f820?, 0x15d0170?})
	/opt/hostedtoolcache/go/1.22.5/x64/src/runtime/panic.go:770 +0x132
github.com/cometbft/cometbft/p2p.(*Switch).AddReactor(0xc0004b0360, {0x1380b23, 0x9}, {0x15e5218, 0xc0004a6070})
	/home/runner/work/cometbft/cometbft/p2p/switch.go:176 +0x27b
github.com/cometbft/cometbft/internal/blocksync.TestNoBlockResponse.func1(0x0, 0xc0004b0360)
	/home/runner/work/cometbft/cometbft/internal/blocksync/reactor_test.go:202 +0x89
github.com/cometbft/cometbft/p2p.MakeSwitch(0xc0002a65a0, 0x0, 0xc0007c3e90, {0x0, 0x0, 0x0})
	/home/runner/work/cometbft/cometbft/p2p/test_util.go:234 +0x984
github.com/cometbft/cometbft/p2p.MakeConnectedSwitches(0xc0002a65a0, 0x2, 0xc0007c3e90, 0x14a63d8)
	/home/runner/work/cometbft/cometbft/p2p/test_util.go:85 +0x95
github.com/cometbft/cometbft/internal/blocksync.TestNoBlockResponse(0xc000450820)
	/home/runner/work/cometbft/cometbft/internal/blocksync/reactor_test.go:201 +0x3cd
testing.tRunner(0xc000450820, 0x14a6250)
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1742 +0x826
```

```
--- FAIL: TestProvider (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15e72d6]

goroutine 19 [running]:
testing.tRunner.func1.2({0x16db280, 0x238ec40})
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1634 +0x6b6
panic({0x16db280?, 0x238ec40?})
	/opt/hostedtoolcache/go/1.22.5/x64/src/runtime/panic.go:770 +0x[132](https://github.com/cometbft/cometbft/actions/runs/10202184809/job/28225738348?pr=3602#step:5:133)
github.com/cometbft/cometbft/node.createSwitch.WithMetrics.func1(0xc0001590e0)
	/home/runner/work/cometbft/cometbft/p2p/switch.go:160 +0x56
github.com/cometbft/cometbft/p2p.NewSwitch(0xc0002e21e0, {0x1b26c28, 0xc000158fc0}, {0xc00005a760, 0x2, 0x20?})
	/home/runner/work/cometbft/cometbft/p2p/switch.go:[133](https://github.com/cometbft/cometbft/actions/runs/10202184809/job/28225738348?pr=3602#step:5:134) +0x7ba
github.com/cometbft/cometbft/node.createSwitch(0xc00046f2c0, {0x1b26c28, 0xc000158fc0}, 0xc000315360, {0x241b6c0, 0x0, 0x0}, {0x1b326e0, 0xc000123200}, {0x1b322b8, ...}, ...)
	/home/runner/work/cometbft/cometbft/node/setup.go:468 +0x1b1
github.com/cometbft/cometbft/node.NewNodeWithCliParams({0x1b26838, 0x241b6c0}, 0xc00046f2c0, {0x1b26a68, 0xc00012d400}, 0xc0001a3780, {0x1b26aa0, 0xc000127a70}, 0xc00005bb18, 0x1984988, ...)
	/home/runner/work/cometbft/cometbft/node/node.go:500 +0x25b5
github.com/cometbft/cometbft/node.NewNode(...)
	/home/runner/work/cometbft/cometbft/node/node.go:291
github.com/cometbft/cometbft/rpc/test.NewCometBFT({0x1b30070, 0xc0001a1e00}, 0xc0001a7b30)
	/home/runner/work/cometbft/cometbft/rpc/test/helpers.go:162 +0x60b
github.com/cometbft/cometbft/rpc/test.StartCometBFT({0x1b30070, 0xc0001a1e00}, {0xc00008fe80, 0x1, 0xc00008fc08?})
	/home/runner/work/cometbft/cometbft/rpc/test/helpers.go:114 +0x105
github.com/cometbft/cometbft/light/provider/http_test.TestProvider(0xc00045b1e0)
	/home/runner/work/cometbft/cometbft/light/provider/http/http_test.go:40 +0x311
testing.tRunner(0xc00045b1e0, 0x1983ed8)
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1742 +0x826
FAIL	github.com/cometbft/cometbft/light/provider/http	0.088s
ok  	github.com/cometbft/cometbft/light/store/db	1.103s	coverage: 70.4% of statements
```

**Comment by cason** [comment]
The `Transport` interface is somehow odd because it is public but relies on a private `peerConfig` field. In fact, it is literally impossible to call the `Transport` methods outside the `p2p` package.

File: p2p/switch.go [hunk 1]
```diff
@@ -137,6 +133,15 @@ func NewSwitch(
 		option(sw)
 	}
 
+	sw.peerConfig = &peerConfig{
```
Comment (cason) [inline p2p/switch.go:136]: You have to update this structure on the `AddReactor` method.

**Comment by cason** [comment]
I would suggest opening an issue and we can take care of this fix.

**Comment by itsdevbear** [comment]
@cason https://github.com/cometbft/cometbft/issues/3639

**Comment by itsdevbear** [comment]
@cason fixed the tests as well, do you still want to merge this?

**Comment by melekes** [comment]
```
==================
WARNING: DATA RACE
Write at 0x00c0002ac[20](https://github.com/cometbft/cometbft/actions/runs/10271558401/job/28421813266?pr=3602#step:6:21)0 by goroutine 74:
  github.com/cometbft/cometbft/p2p.(*MultiplexTransport).Dial()
      /home/runner/work/cometbft/cometbft/p2p/transport.go:236 +0x395
  github.com/cometbft/cometbft/p2p.(*Switch).addOutboundPeerWithConfig()
      /home/runner/work/cometbft/cometbft/p2p/switch.go:727 +0x36e
  github.com/cometbft/cometbft/p2p.(*Switch).DialPeerWithAddress()
      /home/runner/work/cometbft/cometbft/p2p/switch.go:553 +0x1c6
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).dialPeer()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:547 +0x295
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).ensurePeers.func1()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:485 +0x49
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).ensurePeers.gowrap1()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:494 +0x41

Previous write at 0x00c0002ac200 by goroutine 54:
  github.com/cometbft/cometbft/p2p.(*MultiplexTransport).Accept()
      /home/runner/work/cometbft/cometbft/p2p/transport.go:203 +0x1bb
  github.com/cometbft/cometbft/p2p.(*Switch).acceptRoutine()
      /home/runner/work/cometbft/cometbft/p2p/switch.go:631 +0xa1
  github.com/cometbft/cometbft/p2p.(*Switch).OnStart.gowrap1()
      /home/runner/work/cometbft/cometbft/p2p/switch.go:244 +0x33

Goroutine 74 (running) created at:
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).ensurePeers()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:484 +0x824
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).ensurePeersRoutine()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:411 +0x251
  github.com/cometbft/cometbft/p2p/pex.(*Reactor).OnStart.gowrap2()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor.go:151 +0x33

Goroutine 54 (running) created at:
  github.com/cometbft/cometbft/p2p.(*Switch).OnStart()
      /home/runner/work/cometbft/cometbft/p2p/switch.go:244 +0x237
  github.com/cometbft/cometbft/libs/service.(*BaseService).Start()
      /home/runner/work/cometbft/cometbft/libs/service/service.go:146 +0x588
  github.com/cometbft/cometbft/p2p/pex.TestPEXReactorRunning()
      /home/runner/work/cometbft/cometbft/p2p/pex/pex_reactor_test.go:110 +0x44b
  testing.tRunner()
      /opt/hostedtoolcache/go/1.22.5/x64/src/testing/testing.go:1689 +0x[21](https://github.com/cometbft/cometbft/actions/runs/10271558401/job/28421813266?pr=3602#step:6:22)e
  testing.(*T).Run.gowrap1()
      /opt/hostedtoolcache/go/1.[22](https://github.com/cometbft/cometbft/actions/runs/10271558401/job/28421813266?pr=3602#step:6:23).5/x64/src/testing/testing.go:1742 +0x44
==================
```

**Comment by cason** [comment]
So, my general view on this:

1. I am not convinced that producing a new instance of `peerConfig` per transport call is that bad in terms of memory. Of course, if dialing/accepting a peer succeeds we will allocate a connection with them, lots of allocations of course. The profile you provided does not convinced me that is the small `peerConfig` allocation that is producing this memory usage.
2. The number of failing tests indicates that we have some synchronization problems here. Maybe that is the original reason for creating a new `peerConfig` instance per connection attempt: to avoid race conditions, since calls to `Dial` and `Accept` are inherently concurrent.

So, I would proceed with this fix if I become convinced about 1. above. Then we can spend sometime to fix 2.

**Pull request closed without merging**